core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+                yield '%s(%s)' % (rule_spec.rule_type, ','.join(param_list))
+
+    @classmethod
+    def _infer_which_answer_group_and_rule_match_answer(cls, state, rule_str):
+        # First, check whether it matches against the default rule, which
+        # thereby translates to the default outcome.
+        answer_groups = state.interaction.answer_groups
+        if rule_str == cls._DEFAULT_RULESPEC_STR:
+            return (len(answer_groups), 0, None)
+
+        # Otherwise, first RuleSpec instance to match is the winner. The first
+        # pass is to stringify parameters and doing a string comparison. This is
+        # efficient and works for most situations.
+        for answer_group_index, answer_group in enumerate(answer_groups):
+            rule_specs = answer_group.rule_specs
+            for rule_spec_index, rule_spec in enumerate(rule_specs):
+                possible_stringified_rules = list(
+                    cls._stringify_classified_rule(rule_spec))
+                if rule_str in possible_stringified_rules:
+                    return (answer_group_index, rule_spec_index, None)
+
+        # The second attempt involves parsing the rule string and doing an exact
+        # match on the rule parameter values. This needs to be done in the event
+        # that the Python-turned-ascii parameters have their own elements out of
+        # order (such as with a dict parameter).
+        if '(' in rule_str and rule_str[-1] == ')':
+            # http://stackoverflow.com/questions/9623114
+            unordered_lists_equal = lambda x, y: (
+                collections.Counter(
+                    AnswerMigrationJob._get_hashable_value(x)) ==
+                collections.Counter(
+                    AnswerMigrationJob._get_hashable_value(y)))
+
+            paren_index = rule_str.index('(')
+            rule_type = rule_str[:paren_index]
+            param_str_list_str = rule_str[paren_index+1:-1]
+            partial_param_str_list = param_str_list_str.split(',')
+            param_str_list = []
+
+            # Correctly split the parameter list by correcting the results from
+            # naively splitting it by merging subsequent elements in the list if
+            # the comma fell within brackets or parentheses.
+            concat_with_previous = False
+            open_group_count = 0
+            for partial_param in partial_param_str_list:
+                if concat_with_previous:
+                    param_str_list[-1] += ',' + partial_param
+                else:
+                    param_str_list.append(partial_param)
+                for char in partial_param:
+                    if char == '(' or char == '[' or char == '{':
+                        open_group_count += 1
+                    elif char == ')' or char == ']' or char == '}':
+                        open_group_count -= 1
+                concat_with_previous = open_group_count != 0
```

Comment:
Mm -- this may be ok, but what if the parentheses/brackets/etc. fall within quotes, so that they should not contribute to the open_group_count? Would it be better to use something like ast.literal_eval() instead (tacking things onto the end of the current argument until you get something parseable)?

That said, if you don't get an exception below on a trial run then maybe it's ok... but this might cause rules not to match that otherwise should?